### PR TITLE
Complement struct/API doc for HLD

### PR DESCRIPTION
**Pull request by junjiemao1**
This patch series adds the structure and API documentation to header and C
sources in doxygen style and replace the hard-coded API docs with
doxygen-generated ones.

Tracked-On: #1595

### CODE REVIEW

**Comment by junjiemao1** [comment]
@dbkinder @gvancuts Any comment on the doc?

File: hypervisor/arch/x86/io.c [hunk 1]
```diff
@@ -387,11 +416,26 @@ static void empty_io_handler_list(struct vm *vm)
...
 void free_io_emulation_resource(struct vm *vm)
 {
 	empty_io_handler_list(vm);
 }
 
+/**
+ * @brief Allow a VM to access a port I/O range
+ *
+ * This API enables direct access from the given \p vm to the port I/O space
+ * starting from \p address_arg to \p address_arg + \p nbytes - 1.
```
Comment (gvancuts) [inline hypervisor/arch/x86/io.c]: Should we use `port_address` here instead of `address_arg`?

File: hypervisor/arch/x86/io.c [hunk 2]
```diff
@@ -387,11 +416,26 @@ static void empty_io_handler_list(struct vm *vm)
...
+ *
+ * @param vm The VM whose port I/O access permissions is to be changed
+ * @param port_address The starting port address of the range
+ * @param nbytes The size of the range, in byte
```
Comment (gvancuts) [inline hypervisor/arch/x86/io.c]:
```suggestion
 * @param nbytes The size of the range, in bytes
```

File: hypervisor/arch/x86/io.c [hunk 3]
```diff
@@ -387,11 +416,26 @@ static void empty_io_handler_list(struct vm *vm)
...
 }
 
+/**
+ * @brief Allow a VM to access a port I/O range
+ *
+ * This API enables direct access from the given \p vm to the port I/O space
+ * starting from \p address_arg to \p address_arg + \p nbytes - 1.
+ *
+ * @param vm The VM whose port I/O access permissions is to be changed
+ * @param port_address The starting port address of the range
```
Comment (gvancuts) [inline hypervisor/arch/x86/io.c]:
```suggestion
 * @param port_address The start address of the port I/O range
```

File: hypervisor/include/arch/x86/ioreq.h [hunk 4]
```diff
@@ -96,38 +121,194 @@ struct vm_io_handler {
...
+	 * This member is used in pair with \p range_end. See the documentation
+	 * of \p range_end for details.
+	 */
 	uint64_t range_start;
+
+	/**
+	 * @brief The ending address
+	 *
+	 * \p range_start (inclusive) and \p range_end (exclusive) together
+	 * specifies the address range that this handler is expected to
```
Comment (gvancuts) [inline hypervisor/include/arch/x86/ioreq.h]:
```suggestion
	 * specify the address range that this handler is expected to
```

File: hypervisor/include/arch/x86/ioreq.h [hunk 5]
```diff
@@ -96,38 +121,194 @@ struct vm_io_handler {
...
+
+/**
+ * @brief Allow a VM to access a port I/O range
+ *
+ * This API enables direct access from the given \p vm to the port I/O space
+ * starting from \p address_arg to \p address_arg + \p nbytes - 1.
+ *
+ * @param vm The VM whose port I/O access permissions is to be changed
+ * @param port_address The starting port address of the range
+ * @param nbytes The size of the range, in byte
```
Comment (gvancuts) [inline hypervisor/include/arch/x86/ioreq.h]:
```suggestion
 * @param nbytes The size of the range, in bytes
```

File: hypervisor/arch/x86/io.c (hunk 1 above)
Comment (junjiemao1) [reply hypervisor/arch/x86/io.c]: Definitely. Will update.

File: hypervisor/arch/x86/io.c (hunk 2 above)
Comment (junjiemao1) [reply hypervisor/arch/x86/io.c]: Updated.

File: hypervisor/arch/x86/io.c (hunk 3 above)
Comment (junjiemao1) [reply hypervisor/arch/x86/io.c]: Updated. Thanks.

File: hypervisor/include/arch/x86/ioreq.h (hunk 4 above)
Comment (junjiemao1) [reply hypervisor/include/arch/x86/ioreq.h]: Updated. Thanks.

File: hypervisor/include/arch/x86/ioreq.h (hunk 5 above)
Comment (junjiemao1) [reply hypervisor/include/arch/x86/ioreq.h]: Updated. Thanks.

**Comment by junjiemao1** [comment]
@dbkinder I wrapped the text-based diagram in the comments for ``vhm_request`` by ``@verbatim`` so that doxygen will not try formatting them. A problem using ``@verbatim`` is that the stars at the beginning of each line is also included in the block. Does that look better?

**Comment by HuiHuangShi** [comment]
1654<br>No new violations to the coding guideline detected.

**Comment by dbkinder** [comment]
@lijinxia so many PRs/commits trying to make conflicting changes to doc/acrn.doxyfile has created a bit of a mess with duplicate filenames in the INPUT= section.  I'm not going to try to resolve the conflict with the github interface.  Can you help?

**Comment by HuiHuangShi** [comment]
1654<br>No new violations to the coding guideline detected.

File: hypervisor/include/public/acrn_common.h [hunk 6]
```diff
@@ -107,6 +191,8 @@ union vhm_io_request {
  * Based on the rules above, a typical VHM request lifecycle should looks like
  * the following.
  *
+ * @verbatim
```
Comment (dbkinder) [inline hypervisor/include/public/acrn_common.h]: How about this instead:
```
 * Based on the rules above, a typical VHM request lifecycle should look like
 * the following, assuming the initial state is FREE:
 *
 * @verbatim embed:rst:leading-asterisk
 *
 * +-----------------------+-------------------------+----------------------+
 * | SOS vCPU 0            | SOS vCPU x              | UOS vCPU y           |
 * +=======================+=========================+======================+
 * |                       |                         | **Hypervisor**:      |
 * |                       |                         |                      |
 * |                       |                         | - Fill in type,      |
 * |                       |                         |   addr, etc.         |
 * |                       |                         | - Pause UOS vCPU y   |
 * |                       |                         | - Set state to       |
 * |                       |                         |   PENDING **(a)**    |
 * |                       |                         | - Fire upcall to     |
 * |                       |                         |   SOS vCPU 0         |
 * |                       |                         |                      |
 * +-----------------------+-------------------------+----------------------+
 * | **VHM**:              |                         |                      |
 * |                       |                         |                      |
 * | - Scan for pending    |                         |                      |
 * |   requests            |                         |                      |
 * | - Set state to        |                         |                      |
 * |   PROCESSING **(b)**  |                         |                      |
 * | - Assign requests to  |                         |                      |
 * |   clients **(c)**     |                         |                      |
 * |                       |                         |                      |
 * +-----------------------+-------------------------+----------------------+
 * |                       | **Client**:             |                      |
 * |                       |                         |                      |
 * |                       | - Scan for assigned     |                      |
 * |                       |   requests              |                      |
 * |                       | - Handle the            |                      |
 * |                       |   requests **(d)**      |                      |
 * |                       | - Set state to COMPLETE |                      |
 * |                       | - Notify the hypervisor |                      |
 * |                       |                         |                      |
 * +-----------------------+-------------------------+----------------------+
 * |                       | **Hypervisor**:         |                      |
 * |                       |                         |                      |
 * |                       | - resume UOS vCPU y     |                      |
 * |                       |   **(e)**               |                      |
 * |                       |                         |                      |
 * +-----------------------+-------------------------+----------------------+
 * |                       |                         | **Hypervisor**:      |
 * |                       |                         |                      |
 * |                       |                         | - Post-work **(f)**  |
 * |                       |                         | - set state to FREE  |
 * |                       |                         |                      |
 * +-----------------------+-------------------------+----------------------+
 *
 * @endverbatim
```

Comment (dbkinder) [reply hypervisor/include/public/acrn_common.h]: (and need a rebase to get the acrn.doxyfile merge conflicts fixed)

Comment (junjiemao1) [reply hypervisor/include/public/acrn_common.h]: Thanks for the suggestions. I've updated the doc and resolve the conflicts as well.

**Pull request merged**